Improve user object group and role handling

johanjanssens opened the pull request:
This PR closes #383 and implements following changes:

- Allow to get groups by name or id (default)
- Allow to get roles by name or id (default)
- Allow to check if a user has a role by name
- Allow to check if a user is part of a group by name

Both the `KUser::getGroups()`and `KUser::getRoles()` methods have an additional `$by_name`parameter (default is false) which will return the groups or roles by name instead of id.

Both the `KUser:hasGroup()`and `KUser::hasRole()` methods accept group or role names for lookups. If an array is passed it's possible to use both id's or names interchangeably. The code can handle this.

A second `strict`parameter has also been added to both, if true the user needs to be part of all the groups or has all the roles provided. If not the methods will return false.

### Notes: 

- The changes cleanup the work done in https://github.com/joomlatools/joomlatools-framework/pull/276. The `setUser()`method has been removed and the re-implemented as part of the `setData()`method. This gives a cleaner API.

johanjanssens commented:
@ercanozkaya Problem here. The specific Joomla user methods are in ComPagesUser which is the singleton user object. If you load a user through the koowa provider will be created a KUserDefault object.

Two possible ways of fixing this:

- I could add a ComKoowaUserDefault object and implement the methods there too, have the framework return that.  

- I could implement the methods using a trait, would be the first use of a trait in framework, and have ComKoowaUser and ComKoowaDefault both use the trait

ercanozkaya commented on code/libraries/joomlatools/component/koowa/user/user.php:
```diff
@@ -99,15 +103,58 @@ public function getRoles()
+                $roles = $this->getObject('com:koowa.database.table.roles')
+                    ->select(parent::getRoles(), KDatabase::FETCH_ARRAY_LIST);
+
+                $this->__roles = array_map('strtolower', array_column($roles, 'title'));
```
@johanjanssens `array_column` is 5.5 only.

johanjanssens replied:
@ercanozkaya For Pages I'm using PHP7+. This code will not be triggered by our own extensions, so think it's save to use array_column?

johanjanssens replied:
As agreed we will up our fw requirements to PHP 5.6 for 3.4.

johanjanssens commented:
@ercanozkaya @amazeika Would like to get some final feedback from you here.

1. I created a ComKoowaUserDefault object as the previous implementation didn't allow for specialised Koowa user objects.

2. I moved the specific Joomla user logic into a trait so it can be re-used in ComKoowaUser and ComKoowaUserDefault singleton.

3. I refactored some oft the changes made in  https://github.com/joomlatools/joomlatools-framework/pull/276 by @amazeika to make the API cleaner and make the trait just work.

Your eyes and thorough testing would be most appreciated.

amazeika commented:
@johanjanssens Still reviewing the changes ... just wanted to say that when performing groups/roles checks I'm always confontred to two kind of checks: strict (where the user must belong to all groups/roles passed as an argument) and non-strict (in which the test passes if the user belong at least to one usergroup/role). For this I often add a conditional boolean argument named strict. I would consider adding this as in practice this is a check that comes often.

ercanozkaya commented on code/libraries/joomlatools/component/koowa/user/user.php:
```diff
@@ -15,52 +15,49 @@
+     * Set the user data
      *
-     * @return array Koowa user data
+     * @param  array|JUser $user An associative array of data or a JUser object
```
```suggestion
     * @param  array|JUser|Joomla\CMS\User\User $user An associative array of data or a JUser object
```

@johanjanssens Let's add the new one too

johanjanssens replied:
Agreed

ercanozkaya commented on code/libraries/joomlatools/component/koowa/user/user.php:
```diff
@@ -15,52 +15,49 @@
-            'expired'    => !$user->activation,
-            'attributes' => $user->getParameters()->toArray()
-        );
+        if($user instanceof JUser)
```
```suggestion
        if($user instanceof JUser || $user instanceof \Joomla\CMS\User\User)
```

@johanjanssens Just to be sure

johanjanssens replied:
Agreed

johanjanssens commented:
@amazeika Added `'strict`parameter to both hasGroup() and hasRole(). All yours for final review.

@ercanozkaya Have added the check for `\Joomla\CMS\User\User` as requested

ercanozkaya commented:
@johanjanssens  Same here, looks good
